### PR TITLE
avm1: Allow `this` to be `undefined` (fix #843)

### LINKED ISSUE #843: avm1: this can be undefined
Here's a case where this is a pure `undefined`:
```
function f() {
  trace(this);        // undefined
  trace(typeof this); // undefined
}

setTimeout(f, 100);
```

Unfortunately all of our code assumes that `this: Object`.

### PR DESCRIPTION
**Pull request by Herschel**
Change `this` to ~`Option<Object>`~ `Value` in various places to allow for `this` to be undefined. This is caused by:
 * `CallMethod` with an empty method name parameter: `func[""]();`
 * `setInterval` callbacks: `setInterval(func, 100);`

 * I decided to not change `NativeFunction` for now. This could be done in a future PR.
   *  Built-ins will still receive a boxed undefined.
   * This will be incorrect if a built-in does, for example, `this.x = 1; trace(this.x);` which traces 1 with a boxed undefined.
   * We'll probably change how built-ins are defined/generated at some point, and we can change it then.

Fixes #843.

### CODE REVIEW

**Comment by adrian17** [comment]
> this: Value may be better, and would be necessary if there are cases where this can be a primitive. I'm not aware of any way to do this in AVM1.

When calling methods on primitives, like strings?
Long-term I'm thinking this should be done in AVM2 (instead of PrimitiveObject boxing), though maybe it's less relevant for AVM1...

**Comment by Herschel** [comment]
> When calling methods on primitives, like strings?
In AVM1, these get boxed when methods are called on them:
```
String.prototype.foo = function() { trace(this); trace(typeof this); }
"test".foo(); // "test", object
```
But as an optimization we might be able to keep this unboxed for built-ins?

AVM2 doesn't have the concept of boxed primitives from the language side (the above example prints "string"), so I think `this` should definitely be `Value` there.

**Comment by adrian17** [comment]
I see :+1:

**Comment by Herschel** [comment]
It still might be good to do `this: Value<'gc>` just for consistency, and I dread finding the one wacky case where this is actually an unboxed primitive...!

**Comment by Herschel** [comment]
Decided to switch to `this: Value`.

File: core/src/avm1/activation.rs [hunk 1]
```diff
@@ -817,8 +817,7 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
         let result = if method_name.is_empty() {
             // Undefined/empty method name; call `this` as a function.
             // TODO: Pass primitive value instead of boxing (#843).
```
Comment (relrelb) [inline core/src/avm1/activation.rs:819]: This TODO is now done.

**Review by relrelb — APPROVED** [review]

**Pull request merged**